Model/src/main/java/org/gusdb/wdk/model/user/dataset/event/UserDatasetEventSyncHandler.java

```diff
@@ -0,0 +1,277 @@
+import org.gusdb.wdk.model.config.ModelConfig;
+import org.gusdb.wdk.model.user.dataset.UserDataset;
+import org.gusdb.wdk.model.user.dataset.UserDatasetSession;
+import org.gusdb.wdk.model.user.dataset.UserDatasetTypeHandler;
+import org.gusdb.wdk.model.user.dataset.event.model.*;
+
+public class UserDatasetEventSyncHandler extends UserDatasetEventHandler
+{
+  private static final Logger LOG = LogManager.getLogger(UserDatasetEventSyncHandler.class);
+
+  private static final int copyTimeoutMinutes = 180;
+
+  private final HashSet<Long> failedDatasets;
+  private final HashSet<Long> recoveredEvents;
+
+  private final long previousLastHandled;
+
+  private final UserDatasetSession dsSession;
+
+  public UserDatasetEventSyncHandler(
+    UserDatasetSession dsSession,
+    DataSource ds,
+    String dsSchema,
+    String projectId,
+    ModelConfig modelConfig
+  ) {
+    super(ds, dsSchema, projectId, modelConfig);
+
+    this.dsSession  = dsSession;
+
+    failedDatasets      = getEventRepo().getIgnoredDatasetIDs();
+    recoveredEvents     = getEventRepo().getCleanupCompleteEventIDs();
+    previousLastHandled = getEventRepo().getLastHandledEvent();
+  }
+
+  @Override
+  protected boolean attemptEventClaim(EventRow row) {
+    return getEventRepo().claimSyncEvent(row);
+  }
+
+  public void handleShareEvent(UserDatasetShareEvent event) {
+    LOG.info("Updating share of user dataset " + event.getUserDatasetId());
+
+    if (!checkUserDatasetInstalled(event.getUserDatasetId())) {
+      // this can happen if the install was skipped, because the ud was deleted first
+      LOG.info("User dataset " + event.getUserDatasetId() + " is not installed. Skipping share.");
+    } else {
+      if (event.getAction() == UserDatasetShareEvent.ShareAction.GRANT)
+        grantShareAccess(event.getOwnerId(), event.getRecipientId(),
+          event.getUserDatasetId()
+        );
+      else
+        revokeShareAccess(event.getOwnerId(), event.getRecipientId(),
+          event.getUserDatasetId()
+        );
+    }
+    closeEventHandling(event);
+  }
+
+  /**
+   * Checks if the given event should be handled.
+   * <p>
+   * Returns {@code true} if all of the following conditions are met.
+   * <ul>
+   *   <li>The event ID falls after the last completed event OR the event is
+   *   marked as "cleanup_complete".</li>
+   *   <li>The event user dataset ID has no previous failures associated.</li>
+   * </ul>
+   * <p>
+   */
+  @Override
+  public boolean shouldHandleEvent(EventRow row) {
+    // User dataset has a failure associated with it, do not process.
+    if (failedDatasets.contains(row.getUserDatasetID()))
+      return false;
+
+    // Event has been marked as retryable, process if we can lock it.
+    if (recoveredEvents.contains(row.getEventID()))
+      return true;
+
+    // Event is after the last run's last handled event, process if we can lock
+    // it.
+    return row.getEventID() > previousLastHandled;
```

Comment:
@Foxcapades sorry to add new comments after I already made a review but I was thinking about handled_time in the schema. Can you check this is correct?

Say there are installed datasets A, B, and C, and events: "install A" , "install B", and "install C", all completed. I realise B was actually incorrect. I set it to cleanup_ready, run the cleaner, and get it into cleanup_complete. Then I run a sync. UserDatasetEventSync.findLastHandledEvent returns A, because B got cleaned up, and the list of events runs B and C. Then this line will return true for both of them. But C should not be handled - double installing something is not allowed, and it will error out.

Reply:
This shouldn't cause any attempts to reinstall an already installed UD.  `findLastHandledEvent` is just to tell the iRODS component how far back to load events, it doesn't actually mean any of those events will actually be installed.

The logic here is based on the `recoveredEvents` and `previousLastHandled` properties on this class.  `recoveredEvents` is a set of all the events that are in the `CLEANUP_COMPLETE` status, where `lastHandledEvent` is the last event in the `COMPLETED` status.

This means, with your example, that yes, A, B, and C will be pulled from iRODS, but the processing should be:
* A will be ignored (not in `recoveredEvents`, less than `previousLastHandled`
* B will be in the `recoveredEvents` set and thus will be processed
* C will be less than or equal to `previousLastHandled` and will be skipped.

The documentation around how this works should be improved to clarify this.

Reply:
Right, thank you for the extra doc, I was confused about the two uses of the word handled.

```
  public long getLastHandledEvent() {
    var sql = "SELECT MAX(event_id) FROM " + schema + TABLE_USER_DATASET_EVENT
      + " WHERE status = '" + UserDatasetEventStatus.COMPLETE + "'";
```
based on the logic here and the usage later, getLastCompletedEvent might be a better name for this!


I've also realised the handled_time is the first time the event was handled, and doesn't change after the clean-ups. But when I was checking something now I saw there's a doc that says it.